src/compiler/checker.ts

```diff
@@ -13573,23 +13597,30 @@ namespace ts {
          * This is used during inference when instantiating type parameter defaults.
          */
         function createBackreferenceMapper(context: InferenceContext, index: number): TypeMapper {
-            return t => findIndex(context.inferences, info => info.typeParameter === t) >= index ? unknownType : t;
+            return makeFunctionTypeMapper(t => findIndex(context.inferences, info => info.typeParameter === t) >= index ? unknownType : t);
         }
 
         function combineTypeMappers(mapper1: TypeMapper | undefined, mapper2: TypeMapper): TypeMapper;
         function combineTypeMappers(mapper1: TypeMapper, mapper2: TypeMapper | undefined): TypeMapper;
         function combineTypeMappers(mapper1: TypeMapper, mapper2: TypeMapper): TypeMapper {
-            if (!mapper1) return mapper2;
-            if (!mapper2) return mapper1;
-            return t => instantiateType(mapper1(t), mapper2);
+            return !mapper1 ? mapper2 : !mapper2 ? mapper1 : makeCompositeTypeMapper(mapper1, mapper2);
         }
 
-        function createReplacementMapper(source: Type, target: Type, baseMapper: TypeMapper): TypeMapper {
-            return t => t === source ? target : baseMapper(t);
+        function addTypeMapping(mapper: TypeMapper | undefined, source: TypeParameter, target: Type) {
+            return mapper && mapper.kind === TypeMapKind.Simple && mapper.source2 === mapper.target2 ?
+                makeSimpleTypeMapper(mapper.source1, mapper.target1, source, target) :
+                combineTypeMappers(mapper, makeUnaryTypeMapper(source, target));
         }
 
-        function permissiveMapper(type: Type) {
-            return type.flags & TypeFlags.TypeParameter ? wildcardType : type;
+        function createReplacementMapper(source: Type, target: Type, baseMapper: TypeMapper): TypeMapper {
+            switch (baseMapper.kind) {
+                case TypeMapKind.Simple:
+                    return makeSimpleTypeMapper(baseMapper.source1, baseMapper.source1 === source ? target : baseMapper.target1,
+                        baseMapper.source2, baseMapper.source2 === source ? target : baseMapper.target2);
+                case TypeMapKind.Array:
+                    return makeArrayTypeMapper(baseMapper.sources, map(baseMapper.targets, (t, i) => baseMapper.sources[i] === source ? target : t));
+            }
+            return makeFunctionTypeMapper(t => t === source ? target : getMappedType(t, baseMapper));

```

Comment:
Depending on how badly I've mixed up composite maps (see my question above), this seems like it might be equivalent to a composite map with `source-to-target` on the left hand side?

Reply:
No quite. We don't want the second mapping to be applied to the result of the first. We basically just want to replace one of the mappings in the second mapper, which we can do by putting a check in front of the second mapper.

Reply:
How come this transformation (concatenation?) wasn't interesting enough to become a `TypeMapKind`?